source/state_representation/include/state_representation/trajectory/TrajectoryBase.hpp

```diff
@@ -0,0 +1,240 @@
+   * @param new_point the new point
+   */
+  void add_point(const TrajectoryT& new_point);
+
+  /**
+   * @brief Insert new trajectory point between two already existing points
+   * @param new_point the new point
+   * @param pos the desired position of the new point in the queue
+   * @throw std::out_of_range if pos is out of range
+   */
+  void insert_point(const TrajectoryT& new_point, unsigned int pos);
+
+  /**
+   * @brief Get the trajectory point at given index
+   * @param index the index
+   * @return the trajectory point
+   */
+  TrajectoryT& get_point(unsigned int index);
+
+  /**
+   * @brief Set the trajectory point at given index
+   * @param point the new point
+   * @param index the index
+   * @throw std::out_of_range if index is out of range
+   */
+  void set_point(const TrajectoryT& point, unsigned int index);
+
+  /**
+   * @brief Set the trajectory points from a vector of points
+   * @param points vector of new points
+   */
+  void set_points(const std::vector<TrajectoryT>& points);
+
+  /**
+   * @brief Operator overload for returning a single trajectory point and
+   * corresponding time
+   * @param idx the index
+   * @return the trajectory point
+   * @throw std::out_of_range if index is out of range
+   */
+  const TrajectoryT operator[](unsigned int idx) const;
+
+private:
+  std::deque<TrajectoryT> points_;
+};
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase() : State() {
+  this->reset();
+}
+
+template<typename TrajectoryT>
+inline TrajectoryBase<TrajectoryT>::TrajectoryBase(const std::string& name) : State(name) {
+  this->reset();
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::reset() {
+  this->State::reset();
+  this->points_.clear();
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::add_point(const TrajectoryT& new_point) {
+  this->set_empty(false);
+  this->points_.push_back(new_point);
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::insert_point(const TrajectoryT& new_point, unsigned int pos) {
+  if (pos > this->points_.size()) {
+    throw std::out_of_range("Index out of range");
+  }
+  this->set_empty(false);
+  auto it_points = this->points_.begin();
+  std::advance(it_points, pos);
+  this->points_.insert(it_points, new_point);
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::delete_point() {
+  this->set_empty(false);
+  if (!this->points_.empty()) {
+    this->points_.pop_back();
+  }
+}
+
+template<typename TrajectoryT>
+inline void TrajectoryBase<TrajectoryT>::clear() {
+  this->points_.clear();
+}
+
+template<typename TrajectoryT>
+inline const std::deque<TrajectoryT>& TrajectoryBase<TrajectoryT>::get_points() const {
+  return this->points_;
+}
+
+template<typename TrajectoryT>
+inline const TrajectoryT& TrajectoryBase<TrajectoryT>::get_point(unsigned int index) const {
+  return this->points_[index];
```

Comment:
Does this give `std::out_of_range` if invalid index?

Reply:
yes, but this is always checked in derived classes. Having said that, it's better to move it up here